vcsclient/bitbucketcloud.go

```diff
@@ -72,6 +73,61 @@ func (client *BitbucketCloudClient) ListBranches(ctx context.Context, owner, rep
 	return results, nil
 }
 
+// AddSshKeyToRepository on Bitbucket cloud the deploy-key is always read only
+func (client *BitbucketCloudClient) AddSshKeyToRepository(ctx context.Context, owner, repository, keyName, publicKey string, _ Permission) error {
+	// https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Bworkspace%7D/%7Brepo_slug%7D/deploy-keys#post
+	err := validateParametersNotBlank(map[string]string{
+		"owner":      owner,
+		"repository": repository,
+		"key name":   keyName,
+		"public key": publicKey,
+	})
+	if err != nil {
+		return err
+	}
+	endpoint := client.vcsInfo.ApiEndpoint
+	if endpoint == "" {
+		endpoint = bitbucket.DEFAULT_BITBUCKET_API_BASE_URL
+	}
+	u := fmt.Sprintf("%s/repositories/%s/%s/deploy-keys", endpoint, owner, repository)
+	addKeyRequest := bitbucketCloudAddSshKeyRequest{
+		Label: keyName,
+		Key:   publicKey,
+	}
+
+	body := new(bytes.Buffer)
+	err = json.NewEncoder(body).Encode(addKeyRequest)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, u, body)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.SetBasicAuth(client.vcsInfo.Username, client.vcsInfo.Token)
+
+	bitbucketClient := client.buildBitbucketCloudClient(ctx)
+	response, err := bitbucketClient.HttpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = vcsutils.DiscardResponseBody(response)
```

Comment:
What is the reason for consuming the response body before closing it?

Reply:
We ran into a problem once where the connection wasn't closed correctly when the body wasn't consumed.

Reply:
From http.Client.Do() documentation:

If the returned error is nil, the Response will contain a non-nil Body which the user is expected to close. If the Body is not both read to EOF and closed, the Client's underlying RoundTripper (typically Transport) may not be able to re-use a persistent TCP connection to the server for a subsequent "keep-alive" request.